eli5/keras/gradcam.py

```diff
@@ -0,0 +1,188 @@
+        * Kotikalapudi, Raghavendra and contributors for "https://github.com/raghakot/keras-vis".
+    """
+    # For reusability, this function should only use numpy operations
+    # Instead of backend library operations
+    
+    # Perform a weighted linear combination
+    # we need to multiply (dim1, dim2, maps,) by (maps,) over the first two axes
+    # and add each result to (dim1, dim2,) results array
+    # there does not seem to be an easy way to do this:
+    # see: https://stackoverflow.com/questions/30031828/multiply-numpy-ndarray-with-1d-array-along-a-given-axis
+    spatial_shape = activations.shape[:2] # -> (dim1, dim2)
+    lmap = np.zeros(spatial_shape, dtype=np.float64)
+    # iterate through each activation map
+    for i, w in enumerate(weights): 
+        # weight * spatial map
+        # add result to the entire localization map (NOT pixel by pixel)
+        lmap += w * activations[..., i]
+
+    lmap = np.maximum(lmap, 0) # ReLU
+
+    # normalize lmap to [0, 1] ndarray
+    # add eps to avoid division by zero in case lmap is 0's
+    # this also means that lmap max will be slightly less than the 'true' max
+    lmap = lmap / (np.max(lmap)+K.epsilon())
+    return lmap
+
+
+def gradcam_backend(estimator, # type: Model
+    doc, # type: np.ndarray
+    targets, # type: Optional[List[int]]
+    activation_layer # type: Layer
+    ):
+    # type: (...) -> Tuple[np.ndarray, np.ndarray, np.ndarray, int, float]
+    """
+    Compute the terms and by-products required by the Grad-CAM formula.
+    
+    Parameters
+    ----------
+    estimator : keras.models.Model
+        Differentiable network.
+
+    doc : numpy.ndarray
+        Input to the network.
+
+    targets : list, optional
+        Index into the network's output,
+        indicating the output node that will be
+        used as the "loss" during differentiation.
+
+    activation_layer : keras.layers.Layer
+        Keras layer instance to differentiate with respect to.
+    
+
+    See :func:`eli5.keras.explain_prediction` for description of the 
+    ``estimator``, ``doc``, ``targets`` parameters.
+
+    Returns
+    -------
+    (weights, activations, gradients, predicted_idx, predicted_val) : (numpy.ndarray, ..., int, float)
+        Values of variables.
+    """
+    output = estimator.output
+    predicted_idx = _get_target_prediction(targets, estimator)
+    predicted_val = K.gather(output[0,:], predicted_idx) # access value by index
+
+    # output of target layer, i.e. activation maps of a convolutional layer
+    activation_output = activation_layer.output 
+
+    # differentiate ys (scalar) with respect to each of xs (python list of variables)
+    # FIXME: this might have issues in certain cases
+    # See https://github.com/jacobgil/keras-grad-cam/issues/17
+    # a fix is the following piece of code:
+    # grads = [grad if grad is not None else K.zeros_like(var) 
+    #         for (var, grad) in zip([activation_output], grads)]
```

Comment:
I have ran a manual test with VGG16 (the model used in https://github.com/jacobgil/keras-grad-cam/issues/17 that originally broke things) and there seems to be no issue.

https://github.com/jacobgil/keras-grad-cam/issues/17#issuecomment-423057265 and https://github.com/tensorflow/tensorflow/issues/783#issuecomment-175824168 suggest that there is something wrong with the model or the gradient calculation when `grads` contains `None` values. I can rephrase the comment, or raise an exception when this happens to let the user know?

Reply:
I see, thanks for checking!

> raise an exception when this happens to let the user know?

Sounds good. Do you mean adding an extra check somewhere between these two lines?
```
grads, = grads # grads should be a singleton list (because xs is a singleton)
grads =  K.l2_normalize(grads)
```

Reply:
Yeah that sounds good.

https://github.com/TeamHG-Memex/eli5/pull/315/commits/e57c8d696c98314f4755c689b860493a274ccea6